web/scripts/dev/cli.ts

```diff
@@ -0,0 +1,215 @@
+import { performance } from "perf_hooks"
+import formatMS from "pretty-ms"
+import readline from "readline"
+import {
+  answerYesOrNo,
+  error,
+  info,
+  infoline,
+  linebreak,
+  pc,
+  processes,
+  question,
+  section,
+  separator,
+  warn
+} from "../pretty-logs"
+import { Containers } from "./containers"
+import { Mockoon } from "./mockoon"
+import { closing, isBuild, isClean, isServe, pnpm, starting } from "./util"
+import { Vite } from "./vite"
+
+// TODO: add shortcut for rebuilding a container, and then restarting it
+
+export class DevCLI {
+  declare vite?: Vite
+  declare mockoon?: Mockoon
+  declare containers?: Containers
+
+  stopped = false
+  starting = false
+
+  async clean() {
+    if (this.stopped) return
+    this.stopped = true
+
+    // try to stop anything that might still be running
+    for (const proc of processes) {
+      if (!proc.killed) {
+        // we're gonna _REALLY_ try to kill it
+        // `docker-compose` in particular is a pain
+        proc.kill("SIGINT")
```

Comment:
I would suggest waiting a bit after `SIGINT` to see if it's cleaned itself up before proceeding with `SIGKILL`.

Reply:
I can't really tell if the process has ended or not. I have to attach an event listener but that event may never fire. `killed` just indicates if the process has been told to close.